functionary/prompt.py

```diff
@@ -31,147 +187,543 @@ def from_message(cls, message: Dict) -> EndToken:
+        functions = []
+        text_response = None
+        for response in responses:
+            if len(response) == 0:
+                continue
+            # response = assistant<|recipient|>xxxx\n<|content|>yyy
+            recipient_index = response.find(self.recipient_token)
+            content_index = response.find(self.content_token)
+            recipient = response[
+                recipient_index + len(self.recipient_token) : content_index
+            ].strip()
+            content = response[content_index + len(self.content_token) :].strip()
+            # print(f"recipient: {recipient}, content={content}")
+            if recipient == "all":
+                text_response = content
             else:
-                text = f"assistant:\n{StartToken.function.value}{function}:\n{arguments}\n"
-        elif content is not None:  # this is text content
-            text = f"assistant:\n{content}{end_token}\n"
-        else:  # if no function call and content is None --> this is used at inference
-            text = "assistant:"
+                functions.append({"name": recipient, "arguments": content})
 
-    return text
+        tool_calls = []
+        for func in functions:
+            tool_calls.append(
+                {"function": func, "id": get_random_tool_call_id(), "type": "function"}
+            )
+        return {"role": "assistant", "content": text_response, "tool_calls": tool_calls}
 
+    def pre_process_messages_before_inference(self, messages: List[Dict]) -> List[Dict]:
+        """re-order the messages where role = tool to match the order in tool_calls by tool_call_id
+        Args:
+            messages (List[Dict]): list of messages containing: tool_call_id
 
-def convert_old_format_to_openai_format(message: Dict) -> Dict:
-    """convert {"to": xxx, "content": xxx} --> {"function_call": {"name": xxx, "arguments": xxx}}
+        Returns:
+            List[Dict]: _description_
+        """
+        result = []
+        index = 0
+        while index < len(messages):
+            message = messages[index]
+            tool_calls = message.get("tool_calls", None)
+
+            result.append(message)
+            if message["role"] == "assistant" and tool_calls:
+                num_calls = len(tool_calls)
+                if (
+                    tool_calls[0].get("id", None) is not None
+                ):  # if tool_call contains "id" for mapping
+                    tool_call_ids = [item["id"] for item in tool_calls]
+
+                    tool_messages = [messages[index + 1 + j] for j in range(num_calls)]
+                    id_2_tool_messages = {
+                        item["tool_call_id"]: item for item in tool_messages
+                    }
+                    new_messages = [id_2_tool_messages[cid] for cid in tool_call_ids]
+
+                    result.extend(new_messages)
+                    index += num_calls + 1
+                else:
+                    index += 1
```

Comment:
Do we just skip the message if there's no tool_call_id for it?

Reply:
No, we support 2 kinds of formats: with/without tool_call_id:
+ without tool_call_id/id: this is the format of training (you can see that no: id or tool_call_id in in training data). In this format the results of tool calls are already in the correct order.
+ with tool_call_id/id: this is the format for inference, using tool_call_id to put the tool call results in correct order.
For example, tool_calls: [call1-id1; call2-id2]
and the message where role = tool: [{tool_call_id=id2, result2}, {tool_call_id=id1, result1}] --> re-order to: [{tool_call_id=id1, result1},{tool_call_id=id2, result2}]

Reply:
Oh ok, I just realized the logic of this part is that if no id or tool_call_id, just `index += 1` and move on to append the next message so eventually, no messages are skipped.